feat: start DIP0024 from block 1 - fire up test chains by first block - 7/n

knst opened the pull request:
## Issue being fixed or feature implemented
This PR is 7th in the achieving ultimate goal to activate old forks from block 1.
It helps to run unit and functional tests faster; it helps for platform's dev-environment to start faster.


## What was done?
Prior work: #6187, #6189, #6214, #6225, #6269, #6275

This PR:
 - simplify DIP0024 activation and activate it from block 1 at RegTest
 - removes command lines arguments: -bip147height, -dip8params
 - fixed intermittent errors in feature_index_prune.py and feature_llmq_simplepose.py
 - fix assertion crash on Regtest if using strange combination of -testactivationheight and -dip3params

## How Has This Been Tested?
Run unit, functional tests.

## Breaking Changes
[regtest only] -dip8params, -bip147height are removed.


## Checklist:
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have made corresponding changes to the documentation
- [x] I have assigned this pull request to a milestone

knst commented:
converted to draft because `feature_llmq_simplepose.py` is failing often on CI

knst commented:
> converted to draft because feature_llmq_simplepose.py is failing often on CI

CI seems succeed now, please help to review it @UdjinM6 @PastaPastaPasta 

I believe there's still some underlying issue remain in our llmq-code or in our test_framework, because every time something relevant changed, the functional test feature_llmq_simplepose.py becomes unstable:
 - https://github.com/dashpay/dash/pull/6287
 - https://github.com/dashpay/dash/pull/6278
 - https://github.com/dashpay/dash/pull/5859
 - https://github.com/dashpay/dash/pull/3781

knst commented:
> I believe there's still some underlying issue remain in our llmq-code or in our test_framework, because every time something relevant changed, the functional test feature_llmq_simplepose.py becomes unstable:

ok, I fixed intermittent error in feature_llmq_simplepose.py. See https://github.com/dashpay/dash/pull/6325/commits/5ec071a8edc685a9e3e8626844b07339a9e18bcf

What a twist!!!

UdjinM6 commented:
Looks good overall but 45bf3530c6756c60fe5430d13ae923d56e4145d1 and 5092207d1cf1fadabd9b004294364e023e78b966 change nothing for me, tried many times and 1-2 out of 8 jobs are failing with or without these commits. And it feels more like we are hiding the underlying issue rather than fixing it, would prefer them dropped. Also, backport commits should be in their own PR imo.

knst commented:
dropped backports, dropped https://github.com/dashpay/dash/commit/45bf3530c6756c60fe5430d13ae923d56e4145d1 and https://github.com/dashpay/dash/commit/5092207d1cf1fadabd9b004294364e023e78b966 which seems not necessary anymore after [5ec071a](https://github.com/dashpay/dash/pull/6325/commits/5ec071a8edc685a9e3e8626844b07339a9e18bcf)

@UdjinM6 please re-review

UdjinM6 reviewed (dismissed):
utACK f1905ca95039c6d6525483cfaef779a32a419dab

PastaPastaPasta commented on src/evo/creditpool.cpp:
```diff
@@ -114,17 +114,20 @@ void CCreditPoolManager::AddToCache(const uint256& block_hash, int height, const
 static std::optional<CBlock> GetBlockForCreditPool(const CBlockIndex* const block_index, const Consensus::Params& consensusParams)
 {
+    // There's no CbTx before DIP0003 activation
+    if (!DeploymentActiveAt(*block_index, Params().GetConsensus(), Consensus::DEPLOYMENT_DIP0003)) {
```
block_index can be null here?

knst replied:
it seems unrelated because it was used in `ReadBlockFromDisk` before without extra checks:
```
bool ReadBlockFromDisk(CBlock& block, const CBlockIndex* pindex, const Consensus::Params& consensusParams)
{   
    const FlatFilePos block_pos{WITH_LOCK(cs_main, return pindex->GetBlockPos())};
```

Anyway, I added extra annotations just in case, see a new commit

PastaPastaPasta commented on src/llmq/options.cpp:
```diff
@@ -116,26 +116,26 @@ std::map<Consensus::LLMQType, QvvecSyncMode> GetEnabledQuorumVvecSyncEntries()
-    const bool fDIP0024IsActive{optDIP0024IsActive.value_or(DeploymentActiveAfter(pindexPrev, consensusParams, Consensus::DEPLOYMENT_DIP0024))};
-    const bool fHaveDIP0024Quorums{optHaveDIP0024Quorums.value_or(pindexPrev->nHeight >= consensusParams.DIP0024QuorumsHeight)};
+    const bool fDIP0024IsActive{
+        optIsDIP0024Active.value_or(DeploymentActiveAfter(pindexPrev, consensusParams, Consensus::DEPLOYMENT_DIP0024))};
```
nit: we should rework this, currently this will always call DeploymentActiveAfter as it's a param. Should put in a lambda or something to avoid calling DeploymentActiveAfter if optIsDIP0024Active is set

knst replied:
we discussed it, this `optIsDIP0024Active` is used only for Unit test and it means that on production (mainnet, testnet, any devnet) it would be a case of `value_or`. So, this optimization won't give any extra benefits.

Also just a side note (not really important in this particular case) the function `DeploymentActiveAfter` for buried deployment (which is DIP0024) is simplified to one single `if` condition with height comparision, so, there's not much calculation is done anyway

UdjinM6 reviewed (approved):
utACK 906c2d79ba31629cd1637326dc0a6670df78746c

_pls update PR title_

PastaPastaPasta reviewed (approved):
utACK 906c2d79ba31629cd1637326dc0a6670df78746c
